backend/src/datastore/PostDataStore.js

```diff
@@ -0,0 +1,59 @@
+const {DataSource} = require('apollo-datasource');
+
+class PostDataStore extends DataSource {
+
+    constructor(userDataStore, posts = null) {
+        super()
+
+        this.userDataStore = userDataStore;
+        this.posts = posts ?? [
+            {
+                id: 1,
+                title: 'Message 1',
+                votes: [],
+                author_id: 'Max Mustermann'
+            },
+            {
+                id: 2,
+                title: 'Message 2',
+                votes: [
+                    'Max Mustermann'
+                ],
+                author_id: 'Max Mustermann2'
+            }
+
+        ];
+    }
+
+    initialize({context}) {
+    }
+
+    allPosts() {
+        return this.posts;
+    }
+
+    createPost(title, author_id) {
+        let post = {
+            id: Math.max(...this.posts.map(post => post.id), 0) + 1,
+            title: title,
+            votes: [],
+            author_id: author_id
+        }
+        this.posts.push(post);
+        this.userDataStore.createIfNotExists(author_id);
+        return post;
+    }
+
+    upvotePost(id, user) {
+        let post = this.posts.find(post => post.id === parseInt(id));
```

Comment:
> `findIndex((x) => x.id == id)`You can also use this function here, but it will also work fine with indexOf(). This is a bit more reliable.
The find() function is used to find the target element and returns the element if it is found and returns undefined if it is not found.
The findIndex() function also searches for the target element, returns the position of the element if it is found and returns -1 if it is not found.